### PR TITLE
Add docs and screenshots for the debugger plugin

### PR DESCRIPTION
**Pull request by chihuahua**
This is part of #742.

### CODE REVIEW

File: tensorboard/plugins/debugger/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# The Debugger Dashboard
+
+The debugger dashboard offers a graphical user interface for the [TensorFlow debugger](https://www.tensorflow.org/versions/r1.1/programmers_guide/debugger). For instance, this dashboard
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Please use this link instead: https://www.tensorflow.org/programmers_guide/debugger

File: tensorboard/plugins/debugger/README.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+* enables users to pause execution at specified ops or continue execution for some number of steps.
+* visualizes values of tensors.
+* associates tensors with specific lines in python code.
+
+![The Debugger Dashboard](images/debugger_intro.png)
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Can you add and INFO section stressing that we are in the `alpha` release, and not all features may be fully functional yet.

File: tensorboard/plugins/debugger/README.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+# Setup
+
+## Instrumenting Model Logic
+
+To use this dashboard, construct a `TensorBoardDebugWrapperSession`. Subsequently, that wrapper session will issue [gRPC](https://grpc.io/docs/guides/) messages to TensorBoard that containing data for debugging.
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Feel free to defer, but it would be good to point out that for users who use tf-learn's `Estimator`s and `Experiment`s,  `MonitoredSession`s, we provide `TensorBoardDebugHook`. Also point out that there are ways to debug Keras and tf-slim and refer to the aforementioned tfdbg doc link.

File: tensorboard/plugins/debugger/README.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+* enables users to pause execution at specified ops or continue execution for some number of steps.
+* visualizes values of tensors.
+* associates tensors with specific lines in python code.
+
+![The Debugger Dashboard](images/debugger_intro.png)
+
+# Setup
+
+## Instrumenting Model Logic
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: I would move this section after the next two sections, i.e. "## Start TensorBoard with the `debugger_port` flag." and "## Navigating to TensorBoard". That would be the most easy-to-follow order of doing things, because the user will be able to see the initial dialog and how it automatically disappears when the tf.Session connects to tensorboard.

File: tensorboard/plugins/debugger/README.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+Navigate to the debugger dashboard within TensorBoard based on which port it serves on (specified via the `port` flag). For instance, the URL might be `http://localhost:6006#debugger`.
+
+Initially, a dialog may indicate that the dashboard is waiting on a session run to begin. The dialog will hide once the latter happens.
+
+![Dialog on waiting for the first session run](images/debugger_dialog.png)
+
+# Selecting Ops
+
+To examine the output value of an op, ops in the graph must be first selected. Subsequently, the debugger dashboard will pause runs at those ops, enabling users to examine outputs of those ops.
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Strictly speaking, this is not true. Selecting them (i.e., adding breakpoints for them) is one way, but one can do any of the following:
1) right click a node in the graph visualizer and click "continue to"
2) open the source code view, navigate to the relevant line and click the "continue to" (i.e., arrow) icon button. 

So I would rephrase this sentence a little.

File: tensorboard/plugins/debugger/README.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+## The Node List
+
+Ops can be selected via the the node list on the top left:
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Can we try to be consistent in using the words "node" and "op" here? I know the terminology is confusing (even in tensorflow's own documentation), but I tend to refer to the elements the graph as "nodes" (hence the "Runtime Node List". I use the word "op" to refer to a class of "nodes", i.e., the class of all "MatMul" nodes, for example.

File: tensorboard/plugins/debugger/README.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+![Node list](images/node_list.png)
+
+Toggling a checkbox next to an entire scope selects or deselects all ops under the scope. The checkbox for a scope is orange if some but not all of the ops within it are selected.
+
+The ops shown in the list may be filtered by regex based on name or type. Afterwards, ops may be more efficiently selected by the user.
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: -> node name or op type.

File: tensorboard/plugins/debugger/README.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+![Select ops by regex](images/regex_node_list.png)
+
+Clicking the link next to an op makes the graph explorer pan to it (and expand nodes if need be to show the op):
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: pan and zoom

File: tensorboard/plugins/debugger/README.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+![Make the graph explorer pan to a node](images/node_linking.png)
+
+## The Graph Explorer
+
+The graph explorer (on the right side of the dashboard) offers another way to select an op: A context menu appears when an op is right-clicked.
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: This is a somewhat obscure feature. So I suggest  bold-face this word to bring more attention to it.

File: tensorboard/plugins/debugger/README.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+Clicking the button opens a dialog that lets the user specify how many session runs to execute. Execution will pause at breakpoints (selected ops).
+
+![The continue dialog](images/continue_dialog.png)
+
+Via this dialog, the user may specify breakpoints based on conditions (in addition to the breakpoints that are based on selected ops):
+
+* When a tensor contains any bad (NaN, or +/- Infinity) values.
+* When a tensor contains any +/- Infinity values.
+* When a tensor contains any NaN values.
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Add a link to https://en.wikipedia.org/wiki/NaN. Same above.

File: tensorboard/plugins/debugger/README.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+* When the standard deviation of a tensor exceeds some constant.
+* When the standard deviation of a tensor is below some constant.
+
+These conditions bear much semblance to [filters of the TensorFlow debugger](https://www.tensorflow.org/programmers_guide/debugger#finding_nans_and_infs).
+
+When execution is paused, the next op can be stepped to via clicking *Step*. If a program runs multiple sessions, they will be listed within the Session Runs table under the node list.
+
+![Step to the next op](images/step.png)
+
+# Examining Runtime Graphs
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: This section seems to be somewhat redundant given the "## The Graph Explorer" section above. Please remove it or combine them.

File: tensorboard/plugins/debugger/README.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+# Examining the Values of Tensors
+
+When execution is paused, the values of output tensors for all selected ops are shown within the tensor values table (under the graph). The current op is shown in red.
+
+![Tensor value overview](images/tensor_value_overview.png)
+
+Also presented for each op are the name, count (the number of times the op has been executed), data type, and shape.
+
+Next to each op is a *health pill*, which visualizes the proportion of values within the tensor that fall under each of the six categories noted in the legend. A user might use health pills to for instance pinpoint ops that are culprits for producing undesired values (such as NaN).
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Also mention that hovering the mouse cursor above a health pill will open a tooltip which shows more information about the tensor, such as mean and standard deviation.

File: tensorboard/plugins/debugger/README.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+## Tensor Values Visualized
+
+Note the column titled "Value". Clicking to view the value of each op adds a new card (for visualizing the tensor's value) to the Tensor Values pane.
```
Comment (caisq) [inline tensorboard/plugins/debugger/README.md]: Please add a paragraph about the slicing of tensors, maybe something like

> Each tensor-value card provides a input box with which the user can apply [python/numphy-style slicing](https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html) to the tensor value. For example, suppose a tensor has a shape of `(500, 100)`, applying a slicing of `[::5, :2]` will slicing the tensor every 5 indices along the first dimension and take only the first two indices along the second dimension.
> For each tensor, the time axis (i.e., history of the tensor's execution) is treated as an 1D array and python/numpy-style slicing can be applied as well, in an input box of every tensor-value card. For example, the default slicing `-1` selects the latest time point of the tensor's execution, whereas if the user changes it to `:`, the full history of the tensor will be shown, i.e., increasing the rank of the tensor by 1.

File: tensorboard/plugins/debugger/README.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
+# The Debugger Dashboard
+
+The debugger dashboard offers a graphical user interface for the [TensorFlow debugger](https://www.tensorflow.org/versions/r1.1/programmers_guide/debugger). For instance, this dashboard
+
+* enables users to pause execution at specified ops or continue execution for some number of steps.
```
Comment (nfelt) [inline tensorboard/plugins/debugger/README.md]: IMO this list would read slightly better as
> For instance, this dashboard enables users to
> - pause and resume execution at specified ops or numbers of steps
> - visualize values of tensors over time
> - associate tensors with specific lines in python code.

Just a suggestion though.

File: tensorboard/plugins/debugger/README.md (hunk 11 above)
Comment (chihuahua) [reply tensorboard/plugins/debugger/README.md]: Done.

File: tensorboard/plugins/debugger/README.md (hunk 12 above)
Comment (chihuahua) [reply tensorboard/plugins/debugger/README.md]: Done.

File: tensorboard/plugins/debugger/README.md (hunk 13 above)
Comment (chihuahua) [reply tensorboard/plugins/debugger/README.md]: Done.

File: tensorboard/plugins/debugger/README.md (hunk 3 above)
Comment (chihuahua) [reply tensorboard/plugins/debugger/README.md]: Done.

**Comment by caisq** [comment]
@nfelt can we make sure this commit makes it into r1.6?

**Comment by nfelt** [comment]
@caisq yep, will do.
